File: Back-ASPNET/src/RecruitmentAuth.API/Controllers/AuthController.cs
using System.Security.Claims;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using RecruitmentAuth.API.Helpers;
using RecruitmentAuth.Application.DTOs.Auth;
using RecruitmentAuth.Application.Services;

namespace RecruitmentAuth.API.Controllers;

[ApiController]
[Route("api/auth")]
public class AuthController : ControllerBase
{
    private readonly IAuthService _authService;
    private readonly IConfiguration _config;
    private readonly ILogger<AuthController> _logger;

    private const string RefreshTokenCookie = "refresh_token";

    public AuthController(IAuthService authService, IConfiguration config, ILogger<AuthController> logger)
    {
        _authService = authService;
        _config = config;
        _logger = logger;
    }

    private string GetIpAddress() =>
        HttpContext.Connection.RemoteIpAddress?.ToString() ?? "unknown";

    private string? GetUserAgent() =>
        Request.Headers.UserAgent.ToString();

    private string? GetRefreshToken(string? fromBody = null)
    {
        return Request.Cookies[RefreshTokenCookie]
            ?? fromBody;
    }

    private void SetRefreshTokenCookie(string token)
    {
        var expiresInStr = _config["JWT_REFRESH_EXPIRES_IN"] ?? "7d";
        var days = int.TryParse(expiresInStr.TrimEnd('d'), out var d) ? d : 7;

        Response.Cookies.Append(RefreshTokenCookie, token, new CookieOptions
        {
            HttpOnly = true,
            Secure = false, // Set to true in production with HTTPS
            SameSite = SameSiteMode.Lax,
            Expires = DateTimeOffset.UtcNow.AddDays(days),
            Path = "/"
        });
    }

    private void ClearRefreshTokenCookie()
    {
        Response.Cookies.Delete(RefreshTokenCookie, new CookieOptions
        {
            HttpOnly = true,
            Secure = false,
            SameSite = SameSiteMode.Lax,
            Path = "/"
        });
    }

    [HttpGet("csrf-token")]
    public IActionResult GetCsrfToken()
    {
        return Ok(new { success = true, data = new { csrfToken = "" } });
    }

    [HttpPost("register")]
    public async Task<IActionResult> Register([FromBody] RegisterRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, errors = ModelState });

        try
        {
            var result = await _authService.RegisterAsync(request, GetIpAddress(), GetUserAgent());
            SetRefreshTokenCookie(result.RefreshToken);
            return Ok(ApiResponse<AuthResponse>.Ok(result, "Registration successful. Please verify your email."));
        }
        catch (InvalidOperationException ex)
        {
            return Conflict(new { success = false, message = ex.Message });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Register error");
            return StatusCode(500, new { success = false, message = "An error occurred during registration." });
        }
    }

    [HttpPost("login")]
    public async Task<IActionResult> Login([FromBody] LoginRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, errors = ModelState });

        try
        {
            var result = await _authService.LoginAsync(request, GetIpAddress(), GetUserAgent());
            
            if (result.RequiresTwoFactor)
            {
                return Ok(new { success = true, data = new { requiresTwoFactor = true, tempToken = result.TempToken }, message = "Two-factor authentication required" });
            }

            SetRefreshTokenCookie(result.AuthResponse!.RefreshToken);
            return Ok(ApiResponse<AuthResponse>.Ok(result.AuthResponse, "Login successful."));
        }
        catch (UnauthorizedAccessException ex)
        {
            return Unauthorized(new { success = false, message = ex.Message });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Login error");
            return StatusCode(500, new { success = false, message = "An error occurred during login." });
        }
    }

    [HttpPost("2fa/verify")]
    public async Task<IActionResult> VerifyTwoFactor([FromBody] TwoFactorVerifyRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, errors = ModelState });

        try
        {
            var result = await _authService.VerifyTwoFactorAsync(request.TempToken, request.TotpCode, GetIpAddress(), GetUserAgent());
            SetRefreshTokenCookie(result.RefreshToken);
            return Ok(ApiResponse<AuthResponse>.Ok(result, "2FA verification successful."));
        }
        catch (UnauthorizedAccessException ex)
        {
            return Unauthorized(new { success = false, message = ex.Message });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "2FA verify error");
            return StatusCode(500, new { success = false, message = "An error occurred during verification." });
        }
    }

    [HttpPost("oauth/verify-code")]
    public async Task<IActionResult> VerifyOauthCode([FromBody] OauthVerifyRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, errors = ModelState });

        try
        {
            var result = await _authService.VerifyOauthCodeAsync(request.TempToken, request.Code, GetIpAddress(), GetUserAgent());
            SetRefreshTokenCookie(result.RefreshToken);
            return Ok(ApiResponse<AuthResponse>.Ok(result, "OAuth verification successful."));
        }
        catch (UnauthorizedAccessException ex)
        {
            return Unauthorized(new { success = false, message = ex.Message });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "OAuth verify error");
            return StatusCode(500, new { success = false, message = "An error occurred during verification." });
        }
    }

    [Authorize]
    [HttpPost("2fa/setup")]
    public async Task<IActionResult> SetupTwoFactor()
    {
        var userId = User.FindFirstValue(ClaimTypes.NameIdentifier) ?? User.FindFirstValue("sub");
        if (string.IsNullOrEmpty(userId)) return Unauthorized();

        try
        {
            var result = await _authService.SetupTwoFactorAsync(userId);
            return Ok(new { success = true, data = result, message = "Setup initiated." });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { success = false, message = ex.Message });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "2FA setup error");
            return StatusCode(500, new { success = false, message = "An error occurred." });
        }
    }

    [Authorize]
    [HttpPost("2fa/enable")]
    public async Task<IActionResult> EnableTwoFactor([FromBody] TwoFactorEnableRequest request)
    {
        var userId = User.FindFirstValue(ClaimTypes.NameIdentifier) ?? User.FindFirstValue("sub");
        if (string.IsNullOrEmpty(userId)) return Unauthorized();

        try
        {
            var result = await _authService.EnableTwoFactorAsync(userId, request.TotpCode);
            return Ok(new { success = true, data = result, message = "Two-factor authentication enabled." });
        }
        catch (UnauthorizedAccessException ex)
        {
            return Unauthorized(new { success = false, message = ex.Message });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "2FA enable error");
            return StatusCode(500, new { success = false, message = "An error occurred." });
        }
    }

    [Authorize]
    [HttpPost("2fa/disable")]
    public async Task<IActionResult> DisableTwoFactor([FromBody] TwoFactorDisableRequest request)
    {
        var userId = User.FindFirstValue(ClaimTypes.NameIdentifier) ?? User.FindFirstValue("sub");
        if (string.IsNullOrEmpty(userId)) return Unauthorized();

        try
        {
            var result = await _authService.DisableTwoFactorAsync(userId, request.Password);
            return Ok(new { success = true, data = result, message = "Two-factor authentication disabled." });
        }
        catch (UnauthorizedAccessException ex)
        {
            return Unauthorized(new { success = false, message = ex.Message });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { success = false, message = ex.Message });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "2FA disable error");
            return StatusCode(500, new { success = false, message = "An error occurred." });
        }
    }

    [HttpPost("refresh")]
    public async Task<IActionResult> Refresh([FromBody] RefreshTokenRequest? request = null)
    {
        var refreshToken = GetRefreshToken(request?.RefreshToken);
        if (string.IsNullOrEmpty(refreshToken))
            return Unauthorized(new { success = false, message = "Refresh token is required." });

        try
        {
            var result = await _authService.RefreshTokenAsync(refreshToken, GetIpAddress(), GetUserAgent());
            SetRefreshTokenCookie(result.RefreshToken);
            return Ok(ApiResponse<AuthResponse>.Ok(result, "Token refreshed."));
        }
        catch (UnauthorizedAccessException ex)
        {
            ClearRefreshTokenCookie();
            return Unauthorized(new { success = false, message = ex.Message });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Refresh error");
            return StatusCode(500, new { success = false, message = "An error occurred." });
        }
    }

    [HttpPost("logout")]
    public async Task<IActionResult> Logout([FromBody] RefreshTokenRequest? request = null)
    {
        var refreshToken = GetRefreshToken(request?.RefreshToken);
        try
        {
            if (!string.IsNullOrEmpty(refreshToken))
                await _authService.LogoutAsync(refreshToken);

            ClearRefreshTokenCookie();
            return Ok(new { success = true, message = "Logged out successfully." });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Logout error");
            ClearRefreshTokenCookie();
            return Ok(new { success = true, message = "Logged out." });
        }
    }

    [HttpGet("verify-email")]
    public async Task<IActionResult> VerifyEmail([FromQuery] string token)
    {
        if (string.IsNullOrEmpty(token))
            return BadRequest(new { success = false, message = "Token is required." });

        var verified = await _authService.VerifyEmailAsync(token);
        if (!verified)
            return BadRequest(new { success = false, message = "Invalid or expired verification token." });

        var acceptHeader = Request.Headers.Accept.ToString();
        if (acceptHeader.Contains("text/html"))
        {
            var clientUrl = _config["CLIENT_URL"] ?? "http://localhost:4200";
            return Redirect($"{clientUrl}/auth/login?verified=true");
        }

        return Ok(new { success = true, message = "Email verified successfully." });
    }

    [HttpPost("forgot-password")]
    public async Task<IActionResult> ForgotPassword([FromBody] ForgotPasswordRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, errors = ModelState });

        await _authService.ForgotPasswordAsync(request.Email, GetIpAddress());
        return Ok(new { success = true, message = "If that email exists, a reset link has been sent." });
    }

    [HttpPost("reset-password")]
    public async Task<IActionResult> ResetPassword([FromBody] ResetPasswordRequest request)
    {
        if (!ModelState.IsValid)
            return BadRequest(new { success = false, errors = ModelState });

        try
        {
            await _authService.ResetPasswordAsync(request.Token, request.NewPassword);
            return Ok(new { success = true, message = "Password reset successfully." });
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(new { success = false, message = ex.Message });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Reset password error");
            return StatusCode(500, new { success = false, message = "An error occurred." });
        }
    }

    [Authorize]
    [HttpGet("me")]
    public async Task<IActionResult> GetCurrentUser()
    {
        var userId = User.FindFirstValue(ClaimTypes.NameIdentifier)
            ?? User.FindFirstValue("sub");

        if (string.IsNullOrEmpty(userId))
            return Unauthorized(new { success = false, message = "Not authenticated." });

        try
        {
            var user = await _authService.GetCurrentUserAsync(userId);
            return Ok(ApiResponse<UserDto>.Ok(user));
        }
        catch (KeyNotFoundException)
        {
            return NotFound(new { success = false, message = "User not found." });
        }
    }
}

File: Back-laravel/app/Http/Controllers/Auth/AuthController.php
<?php

namespace App\Http\Controllers\Auth;

use App\Helpers\AuditLogHelper;
use App\Helpers\CryptoHelper;
use App\Helpers\ResponseHelper;
use App\Http\Controllers\Controller;
use App\Http\Requests\Auth\ForgotPasswordRequest;
use App\Http\Requests\Auth\LoginRequest;
use App\Http\Requests\Auth\RegisterRequest;
use App\Http\Requests\Auth\ResetPasswordRequest;
use App\Services\AuthService;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Log;

class AuthController extends Controller
{
    public function __construct(private AuthService $authService) {}

    private function setRefreshCookie($refreshToken, $ttlMs)
    {
        return cookie(
            'refreshToken',
            $refreshToken,
            $ttlMs / 60000, // minutes
            '/',
            env('COOKIE_DOMAIN', null),
            env('APP_ENV') === 'production',
            true,
            false,
            'Strict'
        );
    }

    public function register(RegisterRequest $request)
    {
        try {
            $data = $request->validated();
            $result = $this->authService->register($data);
            
            AuditLogHelper::log($request, AuditLogHelper::USER_REGISTERED, null, [
                'email' => $data['email'],
                'role'  => $data['role'] ?? 'CANDIDATE',
            ]);

            return ResponseHelper::created(null, $result['message']);
        } catch (\Throwable $e) {
            Log::error('[AuthController] register', ['error' => $e->getMessage()]);
            return ResponseHelper::error('An internal error occurred. Please try again.', 500);
        }
    }

    public function verifyEmail(Request $request)
    {
        $frontendLogin = env('CLIENT_URL', 'http://localhost:4200') . '/auth/login';
        try {
            $token = $request->query('token');
            if (!$token) {
                return redirect()->away($frontendLogin . '?verify_error=' . urlencode('Token is required'));
            }

            $this->authService->verifyEmail($token);
            
            AuditLogHelper::log($request, AuditLogHelper::EMAIL_VERIFIED);

            return redirect()->away($frontendLogin . '?verified=true');
        } catch (\Throwable $e) {
            $message = $e->getMessage() ?: 'Verification failed. The link may be expired or already used.';
            return redirect()->away($frontendLogin . '?verify_error=' . urlencode($message));
        }
    }

    public function login(LoginRequest $request)
    {
        try {
            $data = $request->validated();
            $ip = $request->ip();
            $ua = substr($request->userAgent() ?? 'unknown', 0, 500);

            $result = $this->authService->login($data, $ip, $ua);

            if (isset($result['requiresTwoFactor'])) {
                return ResponseHelper::success([
                    'requiresTwoFactor' => true,
                    'tempToken'         => $result['tempToken'],
                ], 'Two-factor authentication required');
            }

            AuditLogHelper::log($request, AuditLogHelper::USER_LOGIN);

            return ResponseHelper::success(['accessToken' => $result['accessToken']], 'Login successful')
                ->cookie($this->setRefreshCookie($result['refreshToken'], $result['refreshTtlMs']));
        } catch (\Throwable $e) {
            $status = $e->getCode() ?: 500;
            if ($status < 400 || $status > 599) $status = 500;
            return ResponseHelper::error($e->getMessage(), $status);
        }
    }

    public function refreshToken(Request $request)
    {
        try {
            $rawToken = $request->cookie('refreshToken');

            if (!$rawToken) {
                return ResponseHelper::error('Refresh token not found. Please log in.', 401);
            }

            $ip = $request->ip();
            $ua = substr($request->userAgent() ?? 'unknown', 0, 500);

            $result = $this->authService->refreshToken($rawToken, $ip, $ua);

            AuditLogHelper::log($request, AuditLogHelper::TOKEN_REFRESHED);

            return ResponseHelper::success(['accessToken' => $result['accessToken']], 'Token refreshed')
                ->cookie($this->setRefreshCookie($result['refreshToken'], $result['refreshTtlMs']));
        } catch (\Throwable $e) {
            $status = $e->getCode() ?: 500;
            if ($status < 400 || $status > 599) $status = 500;
            return ResponseHelper::error($e->getMessage(), $status)
                ->withoutCookie('refreshToken');
        }
    }

    public function oauthTokenExchange(Request $request)
    {
        $rawToken = $request->input('refreshToken');

        if (!$rawToken) {
            return ResponseHelper::error('Refresh token is required', 400);
        }

        return ResponseHelper::success(null, 'Token exchanged successfully')
            ->cookie($this->setRefreshCookie($rawToken, env('JWT_REFRESH_TTL', 10080) * 60 * 1000));
    }

    public function verifyOauthCode(Request $request, \App\Services\OAuthService $oauthService)
    {
        $request->validate([
            'tempToken' => 'required|string',
            'code'      => 'required|string|size:6',
        ]);

        try {
            $ip = $request->ip();
            $ua = substr($request->userAgent() ?? 'unknown', 0, 500);

            $result = $oauthService->verifyOauthCode($request->tempToken, $request->code, $ip, $ua);

            AuditLogHelper::log($request, AuditLogHelper::USER_LOGIN);

            return ResponseHelper::success(['accessToken' => $result['accessToken']], 'OAuth verification successful')
                ->cookie($this->setRefreshCookie($result['refreshToken'], $result['refreshTtlMs']));
        } catch (\Throwable $e) {
            $status = $e->getCode() ?: 500;
            if ($status < 400 || $status > 599) $status = 500;
            return ResponseHelper::error($e->getMessage(), $status);
        }
    }

    public function logout(Request $request)
    {
        try {
            $rawToken = $request->cookie('refreshToken');

            if ($rawToken) {
                $this->authService->logout($rawToken);
            }

            AuditLogHelper::log($request, AuditLogHelper::USER_LOGOUT);

            return ResponseHelper::success(null, 'Logged out successfully')
                ->withoutCookie('refreshToken');
        } catch (\Throwable $e) {
            Log::error('[AuthController] logout error', ['error' => $e->getMessage()]);
            return ResponseHelper::success(null, 'Logged out')
                ->withoutCookie('refreshToken');
        }
    }

    public function forgotPassword(ForgotPasswordRequest $request)
    {
        try {
            $data = $request->validated();
            $result = $this->authService->forgotPassword($data['email']);
            
            AuditLogHelper::log($request, AuditLogHelper::PASSWORD_RESET_REQUESTED, null, ['email' => $data['email']]);

            return ResponseHelper::success(null, $result['message']);
        } catch (\Throwable $e) {
            Log::error('[AuthController] forgotPassword error', ['error' => $e->getMessage()]);
            return ResponseHelper::success(null, 'If that email is registered, you will receive a password reset link shortly.');
        }
    }

    public function resetPassword(ResetPasswordRequest $request)
    {
        try {
            $data = $request->validated();
            $result = $this->authService->resetPassword($data['token'], $data['newPassword']);
            
            AuditLogHelper::log($request, AuditLogHelper::PASSWORD_RESET);

            return ResponseHelper::success(null, $result['message']);
        } catch (\Throwable $e) {
            $status = $e->getCode() ?: 500;
            if ($status < 400 || $status > 599) $status = 500;
            return ResponseHelper::error($e->getMessage(), $status);
        }
    }

    public function getCsrfToken(Request $request)
    {
        try {
            // Laravel handles CSRF via Sanctum/VerifyCsrfToken middleware out of the box for web routes.
            // Since this is an API, we can generate a simple custom CSRF token mirroring the custom logic.
            $token = CryptoHelper::generateSecureToken(32);
            $tokenHash = CryptoHelper::hashToken($token);

            return ResponseHelper::success(['csrfToken' => $token], 'CSRF token generated')
                ->cookie('csrfToken', $tokenHash, 120, '/', env('COOKIE_DOMAIN', null), env('APP_ENV') === 'production', false, false, 'Strict');
        } catch (\Throwable $e) {
            Log::error('[AuthController] getCsrfToken', ['error' => $e->getMessage()]);
            return ResponseHelper::error('Internal error', 500);
        }
    }
}
